File: node-server/src/shared/config/load_config.ts
import {Ajv} from 'ajv';
import {Value} from '@sinclair/typebox/value';
import type {Static} from '@sinclair/typebox';
import {NodeEnv, SCHEMA, type ConfigType} from './config_schema.js';
import {configDotenv} from 'dotenv';

/**
 * Loads environment configuration file from privated path
 * Checks that configuration is correctly formatted
 * @param path file path to a dotenv file
 * @returns configuration
 */
export default function loadConfig(path?: string): ConfigType {
  // Load from .env file
  configDotenv({path});

  const ajv = new Ajv({
    allErrors: true,
    useDefaults: true,
    coerceTypes: true,
    allowUnionTypes: true,
  });

  const defaults = Value.Default(SCHEMA, {}) as Static<typeof SCHEMA>;
  const env = Object.assign(defaults, process.env);
  const valid = ajv.validate(SCHEMA, env);
  if (!valid) {
    const error = new Error('Invalid Configuration! ' + ajv.errorsText());
    (error as unknown as {errors: unknown}).errors = ajv.errors;
    throw error;
  }

  // Application configuration object
  const config: ConfigType = Object.freeze({
    nodeEnv: env.NODE_ENV,
    isDevelopment: env.NODE_ENV === NodeEnv.Development,
    isProduction: env.NODE_ENV === NodeEnv.Production,
    log: {
      level: env.LOG_LEVEL,
    },
    server: {
      host: env.HOST,
      port: env.PORT,
      corsOrigin: env.CORS_ORIGIN,
      serverAddress: env.SERVER_ADDRESS,
    },
    whisper: {
      endpoint: env.WHISPER_SERVICE_ENDPOINT,
      reconnectIntervalSec: env.WHISPER_RECONNECT_INTERVAL_SEC,
    },
    auth: {
      required: env.REQUIRE_AUTH,
      sourceToken: env.SOURCE_TOKEN,
      accessTokenBytes: env.ACCESS_TOKEN_BYTES,
      accessTokenRefreshIntervalSec: env.ACCESS_TOKEN_REFRESH_INTERVAL_SEC,
      accessTokenValidPeriodSec: env.ACCESS_TOKEN_VALID_PERIOD_SEC,
      sessionTokenBytes: env.SESSION_TOKEN_BYTES,
      sessionLengthSec: env.SESSION_LENGTH_SEC,
    },
  }) as unknown as ConfigType;

  return config;
}

File: whisper-service/app_config/load_config.py
'''
Helper function to load application configuration

Functions:
    load_config

Classes:
    AppConfig
'''
import os
from dotenv import load_dotenv
from custom_types.config_types import AppConfig


def load_config() -> AppConfig:
    '''
    Loads application config from .env file.

    Returns:
    AppConfig object
    '''
    load_dotenv()

    config = AppConfig()
    config['API_KEY'] = os.environ.get('API_KEY')
    assert len(config['API_KEY']) > 0, 'API_KEY must be non-empty string'

    config['LOG_LEVEL'] = os.environ.get('LOG_LEVEL', 'info')
    assert config['LOG_LEVEL'] in ['trace', 'debug', 'info'], \
        'LOG_LEVEL must be one of: trace, debug, info'

    config['PORT'] = int(os.environ.get('PORT', 8000))
    config['HOST'] = os.environ.get('HOST', '127.0.0.1')

    return config
